Enable trigger logging in webserver

dstandish opened the pull request:
Implemented for all handlers except alibaba.

Different handlers may be implemented in slightly different ways depending on their characteristics.

Blob storage handlers work by writing to file and then uploading when task is complete.  For these handlers, each trigger writes to its own file and at trigger end the file is uploaded with a suffix that distinguishes it from the task log file.

Blob storage handlers include the following:
* GCS
* S3
* WASB
* Alibaba (not implemented due to missing functionality in hook)

For file-based handlers, we have to do two things to make this work with triggers.

The first is, we don't emit to them synchronously.  We pipe the messages through a queue to a QueueListener that runs in a thread and emits them to the potentially-blocking handler.  

The second is, we need to create a distinct instance of the task handler for each trigger because each instance corresponds to a specific file (and each trigger needs to write to a distinct file).  To accomplish this we add a wrapper handler that manages the individual handlers and routes the messages accordingly.

The other category of handlers is a type of "streaming" handler, where messages are pushed through an API to a remote logging service more or less as they are emitted.  This category includes the following handlers:
* Cloudwatch
* Elasticsearch
* Stackdriver

Each of the streaming handlers has slightly different characteristics.

Stackdriver essentially has fully native support for triggers.  It doesn't require wrapping or routing messages through queuelistener.  This is because for one it already runs in a thread and routes messages through a queue internally so as not to be blocking. Additionally it already attached the necessary labels to the record in each call of `emit` so updathing it to generate those based on the task instance attached to the LogRecord was a trivial matter.

Cloudwatch _could_ be made "native" like stackdriver but the underlying library (watchtower, a 3rd party community library) doesn't quite have the necessary behavior. So for now it still requires using the wrapper and queuelistener.

Elasticsearch is a bit of a hybrid.  By default it is filebased and assumes you have a something like filebeat to ship the files.  In  that way it behaves the same as  blob storage handlers. But it has an optional config where messages go to stdout.  It it is possible that it could be modifed such that when run in stdout mode it could be "native" like stackdriver where there's just one instance of the handler but such work is not taken on here.

Try it with sample dag:

```
with DAG(
    dag_id="example_time_delta_sensor_async",
    schedule=None,
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    catchup=False,
    tags=["example"],
):
    wait = TimeDeltaSensorAsync(task_id="wait", delta=datetime.timedelta(seconds=30))
```

Example:

https://user-images.githubusercontent.com/15932138/215254091-bd1555a8-9a83-4938-89b7-afd5cf0591d7.mov

ashb commented on airflow/jobs/triggerer_job.py:
```diff
@@ -368,8 +518,18 @@ async def run_trigger(self, trigger_id, trigger):
 
-    # Main-thread sync API
+            # unsetting ctx_indiv_trigger var restores stdout logging
+            ctx_indiv_trigger.set(None)
```
Rather thatn `set(None)` should we use https://docs.python.org/3/library/contextvars.html#contextvars.ContextVar.reset?

dstandish replied:
i explored that. the thing with that is you need to store a token when you set it and then supply the token when resetting.  and i decided it's not worth the complexity, and moreover, if it _were_ set to something other than None, we wouldn't want to set it back to that.

ashb commented on airflow/utils/log/file_task_handler.py:
```diff
@@ -87,7 +149,14 @@ def close(self):
+        *,
+        is_trigger=False,
+    ) -> str:
+        try_number = try_number + 1 if is_trigger else try_number
```
Ugh :(. I'm not sure there is a better way of doing this (other than entirely refactoring how try_number behaves which is a rather bigger issue than this PR)  but I don't like it

dstandish replied:
yes, sigh

ashb commented on airflow/utils/log/file_task_handler.py:
```diff
@@ -170,27 +247,27 @@ def _read(self, ti: TaskInstance, try_number: int, metadata: dict[str, Any] | No
                              if further calls might get more log text.
                              This is determined by the status of the TaskInstance
-                 log_pos: (absolute) Char position to which the log is retrieved
+                 log_pos: (absolute) Ch ar position to which the log is retrieved
```
```suggestion
                 log_pos: (absolute) Char position to which the log is retrieved
```

ashb commented on airflow/utils/log/file_task_handler.py:
```diff
@@ -253,10 +309,10 @@ def _read(self, ti: TaskInstance, try_number: int, metadata: dict[str, Any] | No
             except Exception as e:
-                log += f"*** Failed to fetch log file from worker. {str(e)}\n"
+                service_to_fetch_from = LogType.TRIGGER if log_type == LogType.TRIGGER else LogType.WORKER
+                log += f"*** Failed to fetch log file from {service_to_fetch_from}. {str(e)}\n"
```
```suggestion
                log += f"*** Failed to fetch log file from {log_type}. {str(e)}\n"
```

I think?

ashb replied:
Or maybe this
```suggestion
                log += f"*** Failed to fetch log file from {log_type or LogType.WORKER}. {str(e)}\n"
```

ashb replied:
Or maybe we set the default to `log_type = LogType.WORKER` instead of None?

dstandish replied:
> Or maybe we set the default to `log_type = LogType.WORKER` instead of None?

probably best

ashb commented on airflow/utils/log/logging_mixin.py:
```diff
@@ -210,7 +210,11 @@ def set_context(logger, value):
-                flag = cast(FileTaskHandler, handler).set_context(value)
+                if is_trigger and getattr(handler, "supports_triggerer", None):
+                    flag = cast(FileTaskHandler, handler).set_context(value, is_trigger=is_trigger)
+                    logger.propagate = False
```
It feels like setting this shouldn't be done here given we have `flag` and `MAINTAIN_PROPAGATE`? (This feels like it would be confusing to have the flag essentially ignored?)

dstandish replied:
i think i might not need this anymore.... lemme see

dstandish replied:
yup, it was a remnant from the heavier-handed approach, where we were setting context on logging mixin in triggerer job... but now we delegate the context setting to the triggerwrapperhandler which sets it on the handler instance directly, and propagation is irrelevant because it's on root logger.

ashb commented on airflow/utils/log/trigger_handler.py:
```diff
@@ -0,0 +1,134 @@
+    :meta private:
+    """
+
+    def __init__(self, level=logging.NOTSET, base_handler=None):
```
This looks like it shouldn't have a default?
```suggestion
    def __init__(self, base_handler, level=logging.NOTSET):
```

dstandish replied:
yeah sounds right

ashb commented on airflow/utils/serve_logs.py:
```diff
@@ -138,19 +138,19 @@ def load(self):
     wsgi_app = create_app()
 
-    worker_log_server_port = conf.getint("logging", "WORKER_LOG_SERVER_PORT")
+    port = port if port is not None else conf.getint("logging", "WORKER_LOG_SERVER_PORT")
```
```suggestion
    port = port or conf.getint("logging", "WORKER_LOG_SERVER_PORT")
```

ashb reviewed (commented):
I'll need to re-read the `configure_trigger_log_handler` a couple of times to grok it I think :lolsob:

dstandish commented on airflow/models/taskinstance.py:
```diff
@@ -547,7 +547,7 @@ def try_number(self):
         """
         # This is designed so that task logs end up in the right file.
-        if self.state in State.running:
+        if self.state == State.RUNNING:
```
@ashb i think this resolves the "log to wrong file" problem.

(so i can remove the try_number + 1 if is_trigger logic i had)

when state is deferred, try number is decremented. so we need the plus one.

dstandish commented on airflow/jobs/triggerer_job.py:
```diff
@@ -17,26 +17,175 @@
+    if TYPE_CHECKING:
+        assert isinstance(task_handler, FileTaskHandler)
+    wrapper_handler = add_handler_wrapper_to_root(task_handler)
+    filter_trigger_logs_from_other_root_handlers(wrapper_handler)
```
@ashb hopefully i made this a little easier to digest by shoving the logic into functions

potiuk replied:
It's quite clear. BTW. This is an interesting problem to solve with blocking I/O in log handlers.

However, I just wonder if this is really the kind of complexity we need here. And yes.I am sorry for being late in the game here as I had other stuff to take a look at, so maybe I do not understand something and maybe I miss a crucial piece of information. Apologies If I do, but I think we can do it slightly simpler, and in the way that we will get full support for any (even completely custom) streaming log handlers from day one.

What would be nice there is that the "Trigger logs" tab would only be needed if you do not use remote logging or when your remote logging does not support streaming during task execution (like s3). In case of streaming loggers instead, we will get the logs interleaved with the regular task logs - which is much better, because then you will see them in one "stream" (barring occasional races which are inevitable in a distributed system - but with proper time synchronisation between nodes and support from log streaming might be limited to whatever precision the logging solution supports (ms/us etc). 

I think for all cases where streaming logs **just works** during code execution, we would not have to do anything - they would be handed out-of-the-box. It would also utilize most of the other code in this PR - just setting up the handlers would be simpler, and it would support all streaming handers out of the box with 0 modifications to them.

The approach I think we could use in this case is as follows:

1) let all trigger code use "airflow.trigger" logger - which would have always a single handler, that would only do non-blocking calls - forwarding the logs to an in-memory queue, including context of the task instance that is the "cause" of the log.

2) Have a separate non-async-io compatible process that reads from the in-memory queue and .... log it to "airflow.task" logger - pretending they are a special "worker". It could also serve the logs in the very same way as the other workers (and handle the Trigger logs tab). No manipulation of handlers is needed in this case. This separate process would have just the same set of loggers any other airflow process has. The only difference is that it would have to pass the context to "set_context" method of the configured logger before writing a log entry. 

3) Either now or in the future this approach could be quite optimized so that the context does not switch too frequently (this is the main problem I see we have when one triggerer handles multiple tasks). For example we should allow N parallel threads (or processes) to write logs - each with its own queue based on hash of the dag_id, task_id, run_id. Then each such thread/process could do its own micro-batching for writing the logs (with slight delay) and sort them according to dag_id, task_id, run_id before writing, so that multiple log entries for the same task run could be written together in a batch with just a single "set_context" call.

The benefit of doing it this way, all (including custom) streaming handlers will be handled out-of-the-box from day 0 - and neither the users nor us will have to make a single change to those handlers to make them work nicely - including the fact that the trigger log entries will be nicely interleaved with the "regular" task logs in case of streaming solutions, which is exactly what our users expect. 

I believe that should be about it. No manipulation with handlers, no code for custom, streaming handlers, we can implement optimisation for context switching, same code handling local logs and serving them as for the streaming solutions.

dstandish replied:
Gave the comment a quick skim will read more closely later but...

I originally interleaved but I thought that separate tab was actually better. Because it's if course easier to see where your messages are coming from this way, and easier to find what you are looking for e.g. trigger logs vs task.  But I could be convinced otherwise.

Re the solution for blocking io. This is a good practice in general and is more user friendly since it makes blocking io in handlers a non problem.  If you Google it is recommended in official python docs.

And see blog post from Martijn Peters I think his name is ...?

Your suggestions seem to describe pretty accurately what I am doing already. Am continuing to try to refine a bit to reduce the adjustments required in existing handlers. E.g. locally I have removed the is trigger param.

Elastic search looks like it may be more tricky

Anyway if you have ideas to simplify we can discuss. Probably ping you on slack next week

dstandish replied:
Another reason I went with separate tabs is, the handlers, in their reader capacity, support streaming, reading from offset. It gets a little weird to think about what that means and how that should work when you have two different streams/files. And then add in that those files could be local, served, or remote. 
After the process is done it's streigforward enough, but your solution has to work live too.
Another part is timestamps... Since we don't necessarily have control over user log format, we might parse incorrectly. 
Neither is necessarily a deal breaker but caused me to explore separate tabs and I was happy with the result so stuck with it.

potiuk replied:
Agree it's very close. 

I think the **only** real difference is the dynamic replacement of the registered handlers and finding the type of them to see if they are compatible or not.

 I **believe** we do not have to do it -  if we make the separate process/thread that  actually directly use already registered handlers - without having to add a wrapper. 

Simply turn the current "wrapper" into 'forwarder" - i.e. the "handler" would only push it to memory queue, and the queue itself would write to regular "airflow.task" without any modifications. 

I have the impression, looking at the code, that the "supports_triggerer" flag is not needed in the handler and if we implement "forwarder" properly, all and every current (streaming) handler of Airflow will work as expected (and the description suggeste it's not the case and more work is needed). 

But maybe I am wrong. and  I do not realize some hidden coupling (i't often when you start actually implementing stuff you find out some hidden couplings implementation details that make simple thing not that simple any more.

Yah. I will discuss on slack .

dstandish replied:
If you have only one handler instance for all triggers how do you ensure that each message goes to its own file?

dstandish replied:
I.e. the messages for each trigger each go to their own files?

dstandish replied:
You'll have to remember that The trigger files must be named differently from the task files because when they're uploaded you don't want to overwrite. And we have to remember that the same trigger could run multiple times on different triggerer instances and so we have to name those separately too so that they're not overwriting each other

dstandish replied:
So the handler has to know that this is a trigger writing not a task. That's what is_trigger is doing right now. But I have simplified that and will push that later.

potiuk replied:
> Another reason I went with separate tabs is, the handlers, in their reader capacity, support streaming, reading from offset. It gets a little weird to think about what that means and how that should work when you have two different streams/files. And then add in that those files could be local, served, or remote.

Yeah. Very good point. I have not though about the "offset" - and I guess it is pretty much "logging-system-dependent" and might get wrong when you have two different processes (or even machines) writing in paralell to the same stream - which indeed has not been the case so far.  So yeah. Agree separate tabs is likely better in this case.

Though I also think (and that's more to @bbovenzi and @pierrejeambrun  - it would be great if we can have the main and triggerer log tab also side-by-side and ideally some way to link scrolling (maybe in the future ) that they could be scrolled through "together". This might get of course extremely tricky and maybe the "linked scrolling" part is not needed for now,  but being able to see the task log and triggerer log at least side-by-side with timestamps so that they can be somehow - even manually-corelllated is I think a must. I think having to switch tabs back-forth to see where you deferred and what happened after is pretty terrible UX.

Also I think still the "find and replace handler" if we can avoid it is way better (I will discuss on slack about it - maybe that is not possible/easy to avoid it but it looks pretty "fishy" to me).

dstandish replied:
The "find and replace" is similar to what we do for tasks already. We find the task handler and move it to root. See task command.py. only difference is here I wrap it so we can instantiate one per trigger so each can go to own file. See also my other pr re k8s exec pod stdout. 

I use the supports flag in part because we don't want to take just any handler and wrap. User may have other custom handlers there for other reasons and we leave that alone for their benefit. And we can't just take any fth cus not all have been updated. Also, I wanted a way that we could conditionally disable the pane and provide good signalling to user when their handler does not support individual triggerer logging, vs when simply no deferral, and having such a flag makes that straightforward.

potiuk replied:
Just posting here the **summary** of our (long) discussion with @dstandish  that we had on slack - if only for the records but also to track and be able to come back to the discussion in case we decide to improve the solution in the future:

* the current approach is a relatively quick way to solve the painful problem of missing trigger logs so it's ok to go for it now and see if more scalable solution is needed

* it has some limitations that we should be aware of (and make our users aware) - the most important is that it creates multiple (1000s or even 10000s) handlers in a single process - each handler having opened file or socket in parallel might lead to exhaution of sockets (especially in environments like docker and k8s where sockets are actually shared resource). Also each Handlers takes some memory (and for example streaming handlers might take more than "small" amount). Also some streaming handlers might not be capable of runnning multiple clients in the same process.

* this might - in effect - limit Triggerer capabilities in terms of number of paralell tasks handled - and those limits will be system/kernel/deployment dependent. Or it might lead to an excessive memory usage by running Triggerer

* we need to extensively test the resources limits/memory consumption and capabilities of streaming handlers to run multiple clients in the same process before releasing this one to see if these anticipated problems are "real". 

* It's possible that even if we find the limits are there, we decide to relase it and state the limits - warn our users and provide them escape hatches (disabling this individual triggerer logging feature, proper diagnosis and helpful advices when we hit the resource limits, advising the users to run multiple triggerers etc.)

Also being constructive - I have a possible improvement propoasl  we can implement later if we decide to merge this one that I think is doable and takes into account the way how triggers work. I will describe it in a separate comment/start a separate thread for that.

uranusjr commented on airflow/utils/log/log_reader.py:
```diff
@@ -34,7 +34,7 @@ class TaskLogReader:
 
     def read_log_chunks(
-        self, ti: TaskInstance, try_number: int | None, metadata
+        self, ti: TaskInstance, try_number: int | None, metadata, log_type=None
```
We should make the new argument keyword-only

dstandish replied:
good call

uranusjr commented on airflow/utils/log/log_reader.py:
```diff
@@ -55,11 +55,14 @@ def read_log_chunks(
         end.
         """
-        logs, metadatas = self.log_handler.read(ti, try_number, metadata=metadata)
+        kwargs = dict(log_type=log_type) if self.supports_triggerer else {}
```
```suggestion
        kwargs = {"log_type": log_type} if self.supports_triggerer else {}
```

Same for other occurrences.

uranusjr commented on airflow/models/trigger.py:
```diff
@@ -87,7 +89,15 @@ def bulk_fetch(cls, ids: Iterable[int], session=None) -> dict[int, Trigger]:
+            .filter(cls.id.in_(ids))
+            .options(joinedload("task_instance"))
+            .options(joinedload("task_instance.trigger"))
+            .options(joinedload("task_instance.trigger.triggerer_job"))
```
Do we only need these for trigger log, and is it possible to not do these for non-trigger log? (And does it make much difference in performance? I’m not even sure about that.)

dstandish replied:
yeah we could do.. i guess the one thing i was fuzzy about is... does it make it an inner join -- here it doesn't matter since the rows are all there... but  i guess we add `isouter=True` and should work i can try if you think good idea.

but i guess it would be marginally more costly ... not sure how much

uranusjr commented on airflow/providers/elasticsearch/log/es_task_handler.py:
```diff
@@ -187,8 +188,18 @@ def _group_logs_by_host(self, logs):
+    @cached_property
+    def supports_triggerer(self):
+        """Not implemented yet."""
+        return False
```
Would it be a good idea to implement this as a class variable instead? A property returning a constant seems awkward.

dstandish replied:
yes it's absolutely awkward. mostly this property is taken care of through inheritance from FTH, but, in this case and in wasb, i had to implement the property because, since they used type annotations on _read, mypy complained about inconsistent signature so i had to add the new kwarg and manually set to not supported (the property looks at signature to tell)

dstandish replied:
but the important point @uranusjr is.... 

if we do class variable, it propagates through inheritance... and this is a no-no because subclasses need to be updated to properly work with triggers.  not huge updates in most cases. and the property on FTH is defined to be false when not yet implemented (excepti my beautiful plan fails when mypy complains)

uranusjr replied:
Don’t properties also propagate through inheritance? I’m obviously missing something…

dstandish replied:
if the property looks at the signature of the instance method, which is always implemented by subclass and has not yet been updated...

dstandish replied:
look at the implementation in FTH, the base class

dstandish replied:
self always refers to instance, so subclasses evaluate it against their methods, not those of the super

dstandish replied:
perhaps you mean that the subclass should implement as attr even though the parent implements as property?

uranusjr replied:
So to sum up

1. Mypy complains if we make this a class var in the parent class but property in children
2. We need this to be a property in children because some of those need runtime evaluation

In that case I guess a property is alright. I wish Mypy can improve on this; it’s actually a quite common pattern. (To achieve this we need a way to declare an _attribute_ as read-only, I digress…)

dstandish replied:
Not quite. Mypy complaints about the signature of _read in the subclass not matching that of the parent because I updated the signature of the parent. This left me with the choice of either removing type annotations from the subclass method (thus avoiding mypy inspection) or updating the subclass method signature even though I am deferring the act of actually updating the subclass to work with triggers. Rather than remove the type annotations, I chose to update the signature. But we use, in this property, inspection of this methods signature to determine whether the subclass supports triggers. Therefore, it would falsely indicate that the subclass does support triggers ( because the subclass signature now has been updated ), therefore we have to implement the property and return false. I did not try just adding it as an attribute because the parent class is a property and I thought it best to stay consistent with respect to that. When the subclass is ultimately updated to support triggers, then we can remove this property from the subclass because the parent class property will then tell the truth.

uranusjr reviewed (commented):
Mostly looks fine to me, some minor thoughts.

uranusjr commented on airflow/jobs/triggerer_job.py:
```diff
@@ -614,7 +614,8 @@ def set_trigger_logging_metadata(self, ti: TaskInstance, trigger_id, trigger):
         """
-        ti.is_trigger_log_context = True
+        if ti:  # can be None in tests
+            ti.is_trigger_log_context = True
```
Which test(s)? We may want to fix them instead

dstandish replied:
Ideally yes but it was a good number of them like 8 or 9 I think (none added in this pr). I would like to leave that for another day. Maybe a to-do would suffice?

uranusjr replied:
Yeah let’s do it in a follow-up PR.

dstandish committed:
Enable trigger logging

fix tests

fix tests

fix tests

remove changes that were not necessary

review comments

last cleanup for removal of is_trigger

simplify

better diff

remove the is_trigger param

simplify

simplify

Add trigger logs web UI

dstandish committed:
Don't emit end_of_log for task deferral

When the task run ends in a deferral, we should neither emit the "end of log" marker nor run the follow-on schedule check.

dstandish committed:
add support for ES

dstandish committed:
get ES working again

dstandish committed:
don't say 'starting' when resuming from deferral

dstandish committed:
fixup! get ES working again

dstandish committed:
make it optional to have logs in sep pane

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -139,6 +139,13 @@
+    """sentinel."""
+
+
+DEFERRED = ArgDeferred()
```
The naming feels a bit weird. Perhaps we could make those functions return `RunResult | None` where `RunResult` is an enum. It would currently has only one member `deferred` but we could potentially use the mechanism for other things.

dstandish replied:
I like it

dstandish replied:
(your suggestion)

ashb commented on airflow/api_connexion/endpoints/log_endpoint.py:
```diff
@@ -84,6 +90,8 @@ def get_log(
             TaskInstance.map_index == map_index,
         )
         .join(TaskInstance.dag_run)
+        .options(joinedload("trigger"))
```
We're always joining against this table, even when we aren't asking for tirgger logs. We probably shouldn't

ashb commented on airflow/api_connexion/openapi/v1.yaml:
```diff
@@ -1656,6 +1656,12 @@ paths:
+      - name: log_type
+        in: query
+        schema:
+          type: string
```
JSONSchema has "enum" types. Does OpenAPI have something similar?

dstandish replied:
i did that initially i can't remember what problems it caused... can try and go back to it

ashb commented on airflow/example_dags/example_time_delta_sensor_async.py:
```diff
@@ -36,6 +36,6 @@
     tags=["example"],
 ) as dag:
-    wait = TimeDeltaSensorAsync(task_id="wait", delta=datetime.timedelta(seconds=10))
+    wait = TimeDeltaSensorAsync(task_id="wait", delta=datetime.timedelta(seconds=30))
```
Unrelated change?

dstandish replied:
sorta, just makes for better testing of this.  and probably good to do anyway for UX

ashb commented on airflow/jobs/triggerer_job.py:
```diff
@@ -17,26 +17,176 @@
+"""
+If this value is true, trigger logging is configured to use TriggerHandlerWrapper
+
+:meta :private
```
```suggestion
:meta private:
```

ashb commented on airflow/jobs/triggerer_job.py:
```diff
@@ -87,19 +241,29 @@ def on_kill(self):
         self.runner.stop = True
 
+    def _kill_listener(self):
+        if self.listener:
```
Do we want to (attempt) to flush anything in the queue before hand?

ashb replied:
Edit: oh, that's what `stop()` does already :+1:

ashb commented on airflow/jobs/triggerer_job.py:
```diff
@@ -87,19 +241,29 @@ def on_kill(self):
             self.runner.stop = True
+            self._kill_listener()
         else:
             self.log.warning("Forcing exit due to second exit signal %s", signum)
```
We don't stop the listner here. How would this play with s3 file uploading?

dstandish replied:
i believe at this point kill lis has already been called. now we're getting a second signal to kill forcibly.  i suspect outcome is uncertain in that case

ashb commented on airflow/jobs/triggerer_job.py:
```diff
@@ -245,11 +410,10 @@ async def arun(self):
             await self.cleanup_finished_triggers()
             # Sleep for a bit
             await asyncio.sleep(1)
-            # Every minute, log status if at least one trigger is running.
```
Why this change? Debug that was left over?

dstandish replied:
it found it preferable to know it was doing stg
and... what's the harm in one line per minute when no other lines... 
is it too much driveby?

potiuk replied:
I like this. It's also a kind of heartbeat.

ashb commented on airflow/utils/log/trigger_handler.py:
```diff
@@ -0,0 +1,134 @@
+
+ctx_task_instance: ContextVar = ContextVar("task_instance")
+ctx_trigger_id: ContextVar = ContextVar("trigger_id")
+ctx_close_handler: ContextVar = ContextVar("close_handler")
```
I think this one isn't used anymore

dstandish replied:
which one? all used i believe

ashb commented on airflow/utils/log/trigger_handler.py:
```diff
@@ -0,0 +1,134 @@
+
+class DropTriggerLogsFilter(logging.Filter):
+    """
+    If record has non-empty attr trigger_id, filter the record.
```
Desc doesn't match up with impl. Do we actually need the separte `indiv_trigger` field? (i.e. if you just had this look at trigger_id I wouldn't have looked twice)

dstandish replied:
what's impl?
the reason not using trigger_id (i considered it) is we can imagine a future where the log record has trigger_id on the record for whatever reason but it's not with this specific meaning and purpose

dstandish replied:
OH ... docstring bad

dstandish committed:
less joinedload

dstandish committed:
docstring

dstandish committed:
use enum for reader mode

uranusjr commented on airflow/utils/log/file_task_handler.py:
```diff
@@ -73,6 +73,19 @@ def _fetch_logs_from_service(url, log_relative_path):
     return response
 
 
+class TriggerLogsPresentationMode(str, Enum):
```
Is the `str` part needed? Most of existing enums in Airflow subclass str for backward compatibility reasons (they were actually str in early versions), but new enums can likely do away with it.

dstandish replied:
i think we could do away, but i think strings may make for friendlier interop with the JS components... wdyt?

uranusjr replied:
In API we can always call `.value` to get the string value

uranusjr replied:
But my opinion is not _too_ strong, especially if we’re looking to potentially move to FastAPI (not sure if this has an effect but the less the possibility is the better)

dstandish replied:
it's a good point, updated

dstandish committed:
use-enum

dstandish committed:
handle odd error

dstandish committed:
enum

dstandish committed:
no extra code

dstandish committed:
add support for cloudwatch

dstandish committed:
add support for gcp

dstandish committed:
--wip-- [skip ci]

dstandish committed:
forward shut down logs arg to task runner

dstandish committed:
ensure that deferral return code propagates to local task job

dstandish committed:
start the process of switching to single pane -- WIP!

dstandish committed:
revert changes to UI

dstandish committed:
s3 ... untested

dstandish committed:
fixups

dstandish committed:
fixups

dstandish committed:
remove some removed stuff

dstandish committed:
fixup remove log type

uranusjr commented on airflow/cli/commands/task_command.py:
```diff
@@ -69,6 +70,18 @@
+    logger = logging.getLogger()
+    h = next((h for h in logger.handlers if hasattr(h, "ctx_task_deferred")), None)
+    if h:
+        h.ctx_task_deferred = True
```
```suggestion
    logger = logging.getLogger()
    with suppress(StopIteration):
        h = next(h for h in logger.handlers if hasattr(h, "ctx_task_deferred"))
        h.ctx_task_deferred = True
```

dstandish replied:
curious, is the rationale here to avoid the "truthy" evaluation on h?

dstandish replied:
or just readability?

uranusjr replied:
Yeah the main thing is truthy evaluation, or any evaluation really. This is probably not an issue for `logger.handlers` in practice, but it’s better to avoid readers needing to consider the posibility entirely.

uranusjr commented on airflow/cli/commands/triggerer_command.py:
```diff
@@ -18,14 +18,36 @@
+    """Starts serve_logs sub-process"""
+    sub_proc = None
+    if skip_serve_logs is False:
+        from functools import partial
```
Should be at the top of the file?

dstandish committed:
Update airflow/cli/commands/task_command.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

dstandish committed:
top import

dstandish committed:
pluggable timestamp parser

dstandish committed:
fix handling when mult messages same timestamp

dstandish committed:
comment

dstandish committed:
no read local if fin

dstandish committed:
no needed

dstandish committed:
not needed

dstandish committed:
docstring

dstandish committed:
docstring

dstandish committed:
support wasb

dstandish committed:
todo

dstandish committed:
fix es eol

dstandish committed:
fix ui logs load bug

dstandish committed:
simplif

dstandish committed:
can disable trigger logging

dstandish committed:
no queues for cloudwatch -- avoids many threads

dstandish committed:
wip -- got "native" trigger handlers working with stackdriver

dstandish committed:
fixup

dstandish committed:
fixup

dstandish committed:
Merge branch 'main' into enable-trigger-logging

dstandish committed:
documentation

dstandish committed:
docs

dstandish committed:
fix some tests

dstandish committed:
fix tests

dstandish committed:
fix static checks

dstandish committed:
fix static checks

dstandish committed:
provider future compat

dstandish committed:
fix tests

uranusjr committed:
Merge SQLAlchemy query options

uranusjr committed:
Simplify is_trigger_log_context declaration

uranusjr committed:
Simplify typing when Path is involved

uranusjr committed:
Log message format

uranusjr committed:
Fix task command tests

uranusjr commented on airflow/executors/base_executor.py:
```diff
@@ -355,14 +355,15 @@ def execute_async(
         raise NotImplementedError()
 
-    def get_task_log(self, ti: TaskInstance, log: str = "") -> None | str | tuple[str, dict[str, bool]]:
+    def get_task_log(self, ti: TaskInstance) -> tuple[list[str], list[str]]:
```
FileTaskHandler has this pattern that would break after this signature change:

```python
response = executor.get_task_log(ti=ti)
if response:
    ...
```

Also need to fix tests that currently use the removed `log` parameter.

dstandish replied:
this bit
```
response = executor.get_task_log(ti=ti)
if response:
    executor_messages, executor_logs = response
```

it's a remnant from when if `get_task_log` not implemented then would return None, but i have since updated baseexecutor implementation to return `[], []` so that bit of defensiveness is not strictly necessary, though it doesn't hurt.

i have fixed tests, hopefully not too many more iterations.

dstandish committed:
cache executor method

dstandish committed:
Merge branch 'main' into enable-trigger-logging

dstandish committed:
stackdriver fix

dstandish committed:
--wip-- [skip ci]

dstandish committed:
Revert "--wip-- [skip ci]"

This reverts commit fdb5d169ca741b63a3179cf02298710158f75b2b.

dstandish committed:
Revert "forward shut down logs arg to task runner"

This reverts commit 3bf09e4aced80f7fe26b8a90fbb3399fde84a5c0.

dstandish committed:
fix tests

dstandish committed:
fix tests

dstandish committed:
fix tests

dstandish committed:
fix tests

dstandish committed:
fix tests

dstandish committed:
fix tests

dstandish committed:
Merge branch 'main' into enable-trigger-logging

Taragolis commented on airflow/jobs/local_task_job.py:
```diff
@@ -203,6 +205,7 @@ def sigusr2_debug_handler(signum, frame):
                     )
         finally:
             self.on_kill()
+            return return_code
```
I guess this return might be the reason why Segmentation Fault handler tests broken, and potentially others like SIGTERM (but their not raise any error or run in separate process).

dstandish committed:
fix tests

dstandish committed:
fix tests

uranusjr reviewed (approved):
Good as far as I can tell

potiuk commented on docs/apache-airflow/administration-and-deployment/logging-monitoring/logging-tasks.rst:
```diff
@@ -47,11 +44,14 @@ These patterns can be adjusted by :ref:`config:logging__log_filename_template`.
+Interleaving of logs
+--------------------
+
+Airflow's remote task logging handlers can broadly be separated into two categories: streaming handlers (such as ElasticSearch, AWS Cloudwatch, and GCP operations logging, formerly stackdriver) and blob storage handlers (e.g. S3, GCS, WASB).
```
I like those descriptions. They nicely explain thed difference between Blob/Streamng handlers (which were a bit implicit so far).

jedcunningham commented on chart/templates/triggerer/triggerer-service.yaml:
```diff
@@ -0,0 +1,48 @@
+################################
+## Airflow triggerer Service
+#################################
+{{- if semverCompare ">=2.2.0" .Values.airflowVersion }}
```
Should this be 2.6.0 instead? There won't be any endpoints in earlier versions?

dstandish replied:
```suggestion
{{- if semverCompare ">=2.6.0" .Values.airflowVersion }}
```

jedcunningham commented on docs/apache-airflow/administration-and-deployment/logging-monitoring/logging-tasks.rst:
```diff
@@ -47,11 +44,14 @@ These patterns can be adjusted by :ref:`config:logging__log_filename_template`.
+
+Airflow's remote task logging handlers can broadly be separated into two categories: streaming handlers (such as ElasticSearch, AWS Cloudwatch, and GCP operations logging, formerly stackdriver) and blob storage handlers (e.g. S3, GCS, WASB).
+
+For blob storage handlers, depending on the state of the task, logs could be in a lot of different places and in multiple different files.  For this reason, we need to check all locations and interleave what we find.  To do this we need to be able to parse the timestamp for each line.  If you are using a custom formatter you may need to override the default parser by providing a callable name at  ...
```
```suggestion
For blob storage handlers, depending on the state of the task, logs could be in a lot of different places and in multiple different files.  For this reason, we need to check all locations and interleave what we find.  To do this we need to be able to parse the timestamp for each line.  If you are using a custom formatter you may need to override the default parser by providing a callable name at Airflow setting ``[logging] interleave_timestamp_parser``.
```

super nit

jedcunningham commented on docs/apache-airflow/administration-and-deployment/logging-monitoring/logging-tasks.rst:
```diff
@@ -122,15 +122,33 @@ When using remote logging, you can configure Airflow to show a link to an extern
+
+.. note:: This is an advanced topic and most users should be able to just use an existing handler from :doc:`apache-airflow-providers:core-extensions/logging`.
+
+In our providers we have a healthy variety of options with all the major cloud providers.  But should you need to implement logging with a different service, and should you then decide to implement a custom FileTaskHandler, there are a few settings to be aware of, particularly in the context of trigger logging.
```
```suggestion
In our providers we have a healthy variety of options with all the major cloud providers.  However, should you need to implement logging with a different service, and should you then decide to implement a custom FileTaskHandler, there are a few settings to be aware of, particularly in the context of trigger logging.
```

jedcunningham commented on docs/apache-airflow/administration-and-deployment/logging-monitoring/logging-tasks.rst:
```diff
@@ -122,15 +122,33 @@ When using remote logging, you can configure Airflow to show a link to an extern
+
+In our providers we have a healthy variety of options with all the major cloud providers.  But should you need to implement logging with a different service, and should you then decide to implement a custom FileTaskHandler, there are a few settings to be aware of, particularly in the context of trigger logging.
+
+Triggers require a shift in the way that logging is set up.  In contrast with tasks, many triggers run in the same process, and with triggers, since they run in asyncio, we have to be mindful of not introducing blocking calls through the logging handler.  And because of the variation in handler behavior (some write to file, some upload to blob storage, some send messages over network as they arri ...
```
```suggestion
Triggers require a shift in the way that logging is set up.  In contrast with tasks, many triggers run in the same process, and with triggers, since they run in asyncio, we have to be mindful of not introducing blocking calls through the logging handler.  Due to the variation in handler behavior (some write to file, some upload to blob storage, some send messages over network as they arrive, some do so in thread), we need to have some way to let triggerer know how to use them.
```

jedcunningham commented on docs/apache-airflow/administration-and-deployment/logging-monitoring/logging-tasks.rst:
```diff
@@ -122,15 +122,33 @@ When using remote logging, you can configure Airflow to show a link to an extern
+
+Triggers require a shift in the way that logging is set up.  In contrast with tasks, many triggers run in the same process, and with triggers, since they run in asyncio, we have to be mindful of not introducing blocking calls through the logging handler.  And because of the variation in handler behavior (some write to file, some upload to blob storage, some send messages over network as they arri ...
+
+To accomplish this we have a few attributes that may be set either on the handler, either instance or the class.  Inheritance is not respected for these parameters, because subclasses of FileTaskHandler may differ from it in the relevant characteristics.  These params are described below:
```
```suggestion
To accomplish this we have a few attributes that may be set on the handler instance or class.  Inheritance is not respected for these parameters, because subclasses of FileTaskHandler may differ from it in the relevant characteristics.  These params are described below:
```

jedcunningham commented on docs/apache-airflow/administration-and-deployment/logging-monitoring/logging-tasks.rst:
```diff
@@ -122,15 +122,33 @@ When using remote logging, you can configure Airflow to show a link to an extern
+
+- ``trigger_should_wrap``: controls whether this handler should be wrapped by TriggerHandlerWrapper.  This is necessary when each instance of handler creates a file handler that it writes all messages to.
+- ``trigger_should_queue``: controls whether the triggerer should put a QueueListener between the event loop and the handler, to ensure blocking IO in the handler does not disrupt the event loop.
+- ``trigger_send_end_marker``: whether an END signal should be sent to the logger when trigger completes. It is used to tell the wrapper to close and remove the individual file handler specific to the trigger that just completed.
```
```suggestion
- ``trigger_send_end_marker``: controls whether an END signal should be sent to the logger when trigger completes. It is used to tell the wrapper to close and remove the individual file handler specific to the trigger that just completed.
```

jedcunningham commented on docs/apache-airflow/administration-and-deployment/logging-monitoring/logging-tasks.rst:
```diff
@@ -122,15 +122,33 @@ When using remote logging, you can configure Airflow to show a link to an extern
+- ``trigger_should_wrap``: controls whether this handler should be wrapped by TriggerHandlerWrapper.  This is necessary when each instance of handler creates a file handler that it writes all messages to.
+- ``trigger_should_queue``: controls whether the triggerer should put a QueueListener between the event loop and the handler, to ensure blocking IO in the handler does not disrupt the event loop.
+- ``trigger_send_end_marker``: whether an END signal should be sent to the logger when trigger completes. It is used to tell the wrapper to close and remove the individual file handler specific to the trigger that just completed.
+- ``trigger_supported``: If ``trigger_should_wrap`` and ``trigger_should_queue`` are not True, we generally assume that the handler does not support triggers.  But if in this case the handler has ``trigger_supported`` set to True, then we'll still move the handler to root at triggerer start so that it will process trigger messages.  Essentially, this should be true for handlers that "natively" su ...
```
```suggestion
- ``trigger_supported``: If ``trigger_should_wrap`` and ``trigger_should_queue`` are not True, we generally assume that the handler does not support triggers.  However, if in this case the handler has ``trigger_supported`` set to True, then we'll still move the handler to root at triggerer start so that it will process trigger messages.  Essentially, this should be true for handlers that "natively" support triggers. One such example of this is the StackdriverTaskHandler.
```

jedcunningham commented on tests/charts/test_extra_env_env_from.py:
```diff
@@ -114,6 +114,7 @@ def test_extra_env(self, k8s_obj_key, env_paths):
                   name: {RELEASE_NAME}-some-secret
             """
         ).lstrip()
+        print(k8s_obj_key)
```
```suggestion
```

jedcunningham reviewed (approved):
Mostly nitpick and grammar. Overall looks good.

dstandish committed:
Update tests/charts/test_extra_env_env_from.py

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
Update docs/apache-airflow/administration-and-deployment/logging-monitoring/logging-tasks.rst

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
Update docs/apache-airflow/administration-and-deployment/logging-monitoring/logging-tasks.rst

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
Update docs/apache-airflow/administration-and-deployment/logging-monitoring/logging-tasks.rst

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
Update docs/apache-airflow/administration-and-deployment/logging-monitoring/logging-tasks.rst

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
Update docs/apache-airflow/administration-and-deployment/logging-monitoring/logging-tasks.rst

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
Update docs/apache-airflow/administration-and-deployment/logging-monitoring/logging-tasks.rst

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

dstandish committed:
Update chart/templates/triggerer/triggerer-service.yaml

jedcunningham commented on chart/templates/triggerer/triggerer-deployment.yaml:
```diff
@@ -44,6 +45,9 @@ metadata:
     {{- toYaml .Values.triggerer.annotations | nindent 4 }}
   {{- end }}
 spec:
+  {{- if $persistence }}
+  serviceName: {{ .Release.Name }}-triggerer
+  {{- end }}
```
I wonder if this specifically also needs to be behind a version flag?

jedcunningham replied:
Can you give it a quick test?

dstandish replied:
you know... what do you think about just keeping it simple with just the 2.2.0 flag on service?  It's not really gonna hurt anything to have the service there for < 2.6 and it's one less bit of logic in the chart

dstandish committed:
restore just some of the original phrasing re however vs but 😛

dstandish committed:
Revert "Update chart/templates/triggerer/triggerer-service.yaml"

This reverts commit 0628fd30c478c9fa61465fbdfd125e96cc094448.

dstandish committed:
Revert "Revert "Update chart/templates/triggerer/triggerer-service.yaml""

This reverts commit 36c7e8c7d4d1bb6a96ab15b6d0d75e4c725c4f82.

dstandish committed:
add ver check for service name

dstandish committed:
version guard networkpolicty

dstandish committed:
format

dstandish committed:
no version guard in triggerer sts

dstandish committed:
no STS in triggerer dep if < 2.6

dstandish committed:
no port if < 2.6

dstandish committed:
no comment

dstandish committed:
template comment

dstandish committed:
Merge branch 'main' into enable-trigger-logging

dstandish committed:
update chart default airflow version

dstandish committed:
Revert "update chart default airflow version"

This reverts commit 611ae28ddfc1e88640ee0a6db11545a144844eea.

dstandish committed:
revert chart tests

dstandish committed:
fix tests

dstandish committed:
fix tests
